File: 01_session/wdi (introduction to cs)/wdi_fifth_list/task3.py
def gcd(n, m):
    if n == 0:
        return m
    if m == 0:
        return n



    result = 1

    while n != m:
        if n < m:
            n, m = m, n

        if n % 2 == 0 and m % 2 == 0:
            n //= 2
            m //= 2
            result *= 2
        elif n % 2 == 0:
            n //= 2
        elif m % 2 == 0:
            m //= 2
        else:
            n -= m


    print(n, result)
    return n*result

print(gcd(15, 125))

File: informatyka [computer science]/[25-S] Algorytmika Praktyczna/list2/task3.cpp
#include <bits/stdc++.h>
using namespace std;

const int MAXN = 200000;
int parentArr[MAXN + 1];
int sizeArr[MAXN + 1];
int result;
int maxSize;


void DSU(int n) {
    for (int i = 1; i <= n; i++) {
        parentArr[i] = i;
        sizeArr[i] = 1;
    }
    result = n;
    maxSize = 1;
}


int findSet(int v) {
    if (parentArr[v] == v) return v;
    return parentArr[v] = findSet(parentArr[v]);
}


void unionSet(int a, int b) {
    a = findSet(a);
    b = findSet(b);
    if (a != b) {
        if (sizeArr[a] < sizeArr[b]) swap(a, b);
        parentArr[b] = a;
        sizeArr[a] += sizeArr[b];
        maxSize = max(maxSize, sizeArr[a]);
        result--;
    }
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    int N, M;
    cin >> N >> M;

    DSU(N);

    for (int i = 0; i < M; i++) {
        int u, v;
        cin >> u >> v;
        unionSet(u, v);
        cout << result << " " << maxSize << "\n";
    }

    return 0;
}
